### PR TITLE
Apple attestation implementation

### PR DESCRIPTION
**Pull request by randomprogramming**
I tried adding support for Apple attestation, but blocked out when I came to the actual parse and validate functions. I added the boilerplate code needed to get this going, what would need to be done is:

1. Finish `appleParseFn` and `appleValidateFn` in [lib/attestations/apple.js](https://github.com/webauthn-open-source/fido2-lib/compare/master...randomprogramming:master#diff-8632a92cbfd1d6c6a6fc069630e64ad54cb33efb3afe9820fa5186b5cb1ddf5e)
2. Implement some tests in [test/parseAppleAttestationData.test.js](https://github.com/webauthn-open-source/fido2-lib/compare/master...randomprogramming:master#diff-a8137799f372990dde4bd841dcb9a71034771f95fcdf87877fec0a5484276694)

I am creating this PR in hopes of someone more skillful picking this up. I will continue to try and get this implemented but any and all help is appreciated.

### CODE REVIEW

**Comment by Hexagon** [comment]
Have a look at 

https://github.com/duo-labs/webauthn/blob/9f1b88ef44cc0e4f5ddf511ed12a3aa468f972d7/protocol/attestation_apple.go

and figure out what each step does, then I guess you can pick up some implementations from

https://github.com/webauthn-open-source/fido2-lib/blob/master/lib/attestations/androidSafetyNet.js or the other existing attestations, to make it happen the fido2-lib-way

If you get stuck, just commit and we'll have a look at it :)

File: lib/attestations/apple.js [hunk 1]
```diff
@@ -1,12 +1,74 @@
...
+  );
+  const extensions = credCert.getExtensions();
+  let attExtBytes;
+  for (let ext of extensions) {
+    if (Array.isArray(ext) && ext.length > 1) {
+      if (ext[0] === "1.2.840.113635.100.8.2") {
+        attExtBytes = ext[1];
+      }
+    }
+  }
```
Comment (randomprogramming) [inline lib/attestations/apple.js:62]: For some reason, the returned extensions are formatted wrongly. Each key of the `extensions` map is an array where the first value is the key, while the second element of the array is what should actually be the value of the map.

File: lib/attestations/apple.js [hunk 2]
```diff
@@ -1,12 +1,74 @@
...
+  if (!attExtBytes) {
+    return false;
+  }
+
+  const decoded = tools.fromBER(attExtBytes);
+  console.log(hash)
+  console.log(nonce);
+  console.log(decoded.result);
+}
```
Comment (randomprogramming) [inline lib/attestations/apple.js:71]: Not sure which of the decoded values here should be compared to the `nonce`, would like some assistance here.

Comment (Hexagon) [reply lib/attestations/apple.js]: Looks like your on the right path here, nonce should equal decoded.result, but the nonce might be some steps down the tree.

Found some reading here, about this exact thing. The example is using a different (lower level) toolset, much of the example code is not needed to get it to work in fido2-lib. But the steps is the same.

https://medium.com/webauthnworks/webauthn-fido2-verifying-apple-anonymous-attestation-5eaff334c849

Also found an actual attestation response, which you could use to build a test case. That would enable me (as a non-apple person) to experiment with this too 😃 

https://github.com/WebauthnWorks/webauthn-snippets/blob/main/attestation-samples/apple-anonymous.json

Comment (randomprogramming) [reply lib/attestations/apple.js]: Sorry I should've done that sooner, but I added a test with apple attestation so you can check it out now

Comment (Hexagon) [reply lib/attestations/apple.js]: Great 👍

**Comment by Hexagon** [comment]
@randomprogramming Sent a PR for step 4 and 5

**Comment by randomprogramming** [comment]
`TypeError: Key must be of type KeyObject. Received an instance of CryptoKey` Any leads as to what is causing this?

**Comment by Hexagon** [comment]
> `TypeError: Key must be of type KeyObject. Received an instance of CryptoKey` Any leads as to what is causing this?

Which version of node are you running, i've seen that before... Somewhere

**Comment by randomprogramming** [comment]
> > `TypeError: Key must be of type KeyObject. Received an instance of CryptoKey` Any leads as to what is causing this?
> 
> Which version of node are you running, i've seen that before... Somewhere

v14.17.5 on my Macbook where that error appears, v16.14.2 on my PC and that error doesn't appear

**Comment by Hexagon** [comment]
Ah, I'll switch to 14 and have a look. The differen e is that in Node 14 @peculiar/webcrypto is used, while in 16 native webcrypto is used.

**Comment by Hexagon** [comment]
@randomprogramming New pull request sent, fixing Node 14

**Comment by randomprogramming** [comment]
Thank you so much for the help @Hexagon . Did some testing using touchid on the macbook and it seems to be handling it well. Need to try faceid still.

**Comment by Hexagon** [comment]
Cool! Covert this from Draft to PR and drop a note  when you are ready for reviews 👍

**Comment by randomprogramming** [comment]
@Hexagon Would you be able to take a look at the conflicts? Seems to be the code you committed

**Comment by Hexagon** [comment]
Not sure why i couldn't merge this properly, but merge my PR to your master, and open a new PR on fido2-lib. And we should be goot to go again 👍

**Pull request closed without merging**